File: 20-actors/matsurigoto/methods/modules/test_tax_assess.py
#!/usr/bin/env python3
"""Conformance tests for the tax-assess module (matsurigoto 政, ADR-2606052300).

The income-tax assertions reproduce the published JP 速算表 (quick-calc table) EXACTLY —
that is the conformance anchor proving the engine computes real progressive tax, not a toy.
JP quick formula: liability = taxable × marginal_rate − deduction_constant.

Standalone-runnable AND pytest-compatible.
"""
from __future__ import annotations

import pathlib
import sys

HERE = pathlib.Path(__file__).resolve().parent
sys.path.insert(0, str(HERE))

import tax_assess as T  # noqa: E402


def test_no_server_authority():
    """G1 — the module holds no signing authority and signs nothing."""
    assert T.SERVER_HELD_AUTHORITY is False
    r = T.assess_from_return(5_000_000, 0, "JPN.income")
    assert r["receipt"]["proof"] is None
    assert r["receipt"]["server_held_authority"] is False


def test_jp_quick_table_5m():
    """taxable 5,000,000 → 5,000,000×0.20 − 427,500 = 572,500 (JP 速算表)."""
    r = T.assess_income_tax(5_000_000, T.RATE_TABLES["JPN.income"]["brackets"])
    assert r["liability"] == 572_500.0


def test_jp_quick_table_3m():
    """taxable 3,000,000 → 3,000,000×0.10 − 97,500 = 202,500."""
    r = T.assess_income_tax(3_000_000, T.RATE_TABLES["JPN.income"]["brackets"])
    assert r["liability"] == 202_500.0


def test_jp_quick_table_20m():
    """taxable 20,000,000 → 20,000,000×0.40 − 2,796,000 = 5,204,000."""
    r = T.assess_income_tax(20_000_000, T.RATE_TABLES["JPN.income"]["brackets"])
    assert r["liability"] == 5_204_000.0


def test_jp_top_bracket_50m():
    """taxable 50,000,000 → 50,000,000×0.45 − 4,796,000 = 17,704,000 (top 45% bracket)."""
    r = T.assess_income_tax(50_000_000, T.RATE_TABLES["JPN.income"]["brackets"])
    assert r["liability"] == 17_704_000.0


def test_zero_income_zero_tax():
    r = T.assess_income_tax(0, T.RATE_TABLES["JPN.income"]["brackets"])
    assert r["liability"] == 0.0
    assert r["effective_rate"] == 0.0


def test_negative_income_raises():
    try:
        T.assess_income_tax(-1, T.RATE_TABLES["JPN.income"]["brackets"])
    except ValueError:
        return
    raise AssertionError("negative income must raise")


def test_flat_rate_localization():
    """Localization generality: a flat-20% table gives 20% of taxable."""
    r = T.assess_income_tax(1_000_000, T.RATE_TABLES["FLAT20.income"]["brackets"])
    assert r["liability"] == 200_000.0
    assert r["effective_rate"] == 0.20


def test_deductions_reduce_taxable():
    r = T.assess_from_return(gross_income=6_000_000, deductions=1_000_000, table_key="JPN.income")
    assert r["taxable_income"] == 5_000_000
    assert r["liability"] == 572_500.0
    assert r["currency"] == "JPY"


def test_deductions_floor_at_zero():
    r = T.assess_from_return(gross_income=500_000, deductions=900_000, table_key="JPN.income")
    assert r["taxable_income"] == 0.0
    assert r["liability"] == 0.0


def test_unknown_table_raises():
    try:
        T.assess_from_return(1, 0, "NOPE.income")
    except ValueError:
        return
    raise AssertionError("unknown table must raise")


def test_vat_net_due_and_refund():
    due = T.assess_vat(output_vat=300_000, input_vat=120_000, currency="JPY")
    assert due["net_vat_due"] == 180_000.0
    assert due["refund_due"] == 0.0
    refund = T.assess_vat(output_vat=100_000, input_vat=160_000, currency="JPY")
    assert refund["net_vat_due"] == 0.0
    assert refund["refund_due"] == 60_000.0


def test_effective_rate_is_below_top_marginal():
    """Progressive: effective rate < top marginal rate (sanity on the bracket walk)."""
    r = T.assess_income_tax(20_000_000, T.RATE_TABLES["JPN.income"]["brackets"])
    assert r["effective_rate"] < 0.40


def test_solve_is_gated_at_r0():
    try:
        T.solve()
    except RuntimeError:
        return
    raise AssertionError("solve() must raise at R0 (live filing is gated)")


def test_r1d_rate_tables_loaded_for_each_country():
    """R1.D: per-jurisdiction tables load from data/rates/*.edn."""
    for key in ("JPN.income", "USA.income", "DEU.income", "GBR.income", "KOR.income", "IND.income"):
        assert key in T.RATE_TABLES, key
        assert T.RATE_TABLES[key]["brackets"], key


def test_usa_lowest_bracket_10pct():
    r = T.assess_from_return(10_000, 0, "USA.income")
    assert r["liability"] == 1_000.0           # 10% of 10,000
    assert r["currency"] == "USD"


def test_gbr_personal_allowance_zero_tax():
    r = T.assess_from_return(10_000, 0, "GBR.income")
    assert r["liability"] == 0.0               # below the £12,570 allowance
    assert r["currency"] == "GBP"


def test_ind_new_regime_below_threshold_zero():
    r = T.assess_from_return(250_000, 0, "IND.income")
    assert r["liability"] == 0.0               # below ₹300,000


def test_kor_currency_and_progression():
    r = T.assess_from_return(20_000_000, 0, "KOR.income")
    assert r["currency"] == "KRW"
    assert r["effective_rate"] < 0.15          # progressive, below 2nd marginal


def _run():
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_") and callable(v)]
    for fn in fns:
        fn()
        print(f"  ok  {fn.__name__}")
    print(f"\n{len(fns)}/{len(fns)} passed")
    return 0


if __name__ == "__main__":
    sys.exit(_run())

File: 20-actors/matsurigoto/methods/modules/test_tax_assess.clj
;; test_tax_assess.clj — matsurigoto tax-assess conformance + byte-equivalent parity with
;; tax_assess.py against the published JP 速算表. Run via `bb test:matsurigoto`. ADR-2606142300.
(ns matsurigoto.methods.modules.test-tax-assess
  (:require [clojure.test :refer [deftest is testing run-tests]]
            [matsurigoto.methods.modules.tax-assess :as ta]))

(def ^:private jp (get-in ta/embedded-rate-tables ["JPN.income" :brackets]))

(deftest jp-quick-calc-table-parity
  (testing "JP 速算表 reference liabilities reproduced exactly (golden from tax_assess.py)"
    (is (== 0.0       (:liability (ta/assess-income-tax 0 jp))))
    (is (== 97500.0   (:liability (ta/assess-income-tax 1950000 jp))))
    (is (== 572500.0  (:liability (ta/assess-income-tax 5000000 jp))))
    (is (== 1434000.0 (:liability (ta/assess-income-tax 9000000 jp))))
    (is (== 5204000.0 (:liability (ta/assess-income-tax 20000000 jp))))
    (is (== 17704000.0 (:liability (ta/assess-income-tax 50000000 jp))))))

(deftest effective-rate-half-even
  (testing "effective rate (round 6, half-to-even) matches Python golden"
    (is (== 0.0      (:effective-rate (ta/assess-income-tax 0 jp))))
    (is (== 0.05     (:effective-rate (ta/assess-income-tax 1950000 jp))))
    (is (== 0.1145   (:effective-rate (ta/assess-income-tax 5000000 jp))))
    (is (== 0.159333 (:effective-rate (ta/assess-income-tax 9000000 jp))))
    (is (== 0.2602   (:effective-rate (ta/assess-income-tax 20000000 jp))))
    (is (== 0.35408  (:effective-rate (ta/assess-income-tax 50000000 jp))))))

(deftest bracket-breakdown
  (testing "per-bracket breakdown for taxable 5,000,000 (golden)"
    (let [bs (:brackets (ta/assess-income-tax 5000000 jp))]
      (is (= 3 (count bs)))
      (is (= [97500.0 135000.0 340000.0] (mapv #(double (:tax-in-bracket %)) bs)))
      (is (= [0 1950000 3300000] (mapv :lower bs)))
      (is (= [0.05 0.10 0.20] (mapv :rate bs))))))

(deftest assess-from-return-and-receipt
  (testing "gross−deductions → taxable, with G1 unsigned receipt"
    (let [r (ta/assess-from-return 6000000 1000000 "JPN.income")]
      (is (== 572500.0 (:liability r)))
      (is (= "JPY" (:currency r)))
      (is (= "JPN.income" (:rate-table r)))
      (let [rc (:receipt r)]
        (is (nil? (:proof rc)))                          ; G1 — signs nothing
        (is (false? (:server-held-authority rc)))         ; G1
        (is (= "assessed-unsigned" (:status rc)))
        (is (== 572500.0 (:assessed-amount rc)))))))

(deftest vat-net-and-refund
  (testing "net VAT = output − input; negative → refund (G1 unsigned receipt)"
    (let [due (ta/assess-vat 500.50 300.25 "JPY")
          ref (ta/assess-vat 300.0 500.0 "JPY")]
      (is (== 200.25 (:net-vat-due due)))
      (is (== 0.0 (:refund-due due)))
      (is (== 0.0 (:net-vat-due ref)))
      (is (== 200.0 (:refund-due ref)))
      (is (false? (:server-held-authority (:receipt due)))))))

(deftest localization-rate-tables
  (testing "load-rate-tables! merges data/rates/*.edn (universal algorithm, localized param)"
    (ta/load-rate-tables!)
    (is (contains? @ta/rate-tables "USA.income"))
    (is (= "USD" (get-in @ta/rate-tables ["USA.income" :currency])))
    (is (>= (count @ta/rate-tables) 7))                  ; JPN+FLAT20 embedded + ≥5 jurisdictions
    (testing "USA federal single-filer brackets assess (localization generality)"
      (is (== 0.0 (:liability (ta/assess-income-tax 0 (get-in @ta/rate-tables ["USA.income" :brackets]))))))))

(deftest discipline-and-guards
  (testing "input guards + G1 live-filing gate"
    (is (thrown? Exception (ta/assess-income-tax -1 jp)))     ; negative income
    (is (thrown? Exception (ta/assess-income-tax 1000 [])))   ; empty brackets
    (is (thrown? Exception (ta/assess-from-return 1 0 "NOPE.income")))  ; unknown table
    (is (thrown? Exception (ta/solve)))                       ; live filing Council+operator gated
    (is (false? ta/server-held-authority))))                  ; G1

(when (= *file* (System/getProperty "babashka.file"))
  (let [{:keys [fail error]} (run-tests 'matsurigoto.methods.modules.test-tax-assess)]
    (System/exit (if (pos? (+ fail error)) 1 0))))
